File: articles/urban-systems-congestion-housing-and-infrastructure/python/run_all_urban_workflows.py
#!/usr/bin/env python3
"""Run all dependency-light Python workflows for the urban systems article."""

from __future__ import annotations

from pathlib import Path
import subprocess
import sys

ROOT = Path(__file__).resolve().parents[1]
SCRIPTS = [
    "urban_systems_model.py",
    "validation_checks.py",
    "congestion_induced_demand_scenarios.py",
    "housing_transport_affordability.py",
    "infrastructure_maintenance_model.py",
    "displacement_risk_diagnostics.py",
    "urban_resilience_index.py",
    "export_urban_outputs.py",
]


def main() -> None:
    for script in SCRIPTS:
        print(f"\n=== Running {script} ===")
        subprocess.run([sys.executable, str(ROOT / "python" / script)], cwd=str(ROOT), check=True)
    print("\nAll dependency-light Python urban systems workflows completed.")


if __name__ == "__main__":
    main()

File: articles/urban-systems-congestion-housing-and-infrastructure/r/run_all_urban_workflows.R
# Run all base R urban systems workflows with safe path handling.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
if (length(file_arg) > 0) {
  script_path <- normalizePath(sub("^--file=", "", file_arg[1]), mustWork = FALSE)
  root_dir <- normalizePath(file.path(dirname(script_path), ".."), mustWork = FALSE)
} else {
  cwd <- normalizePath(getwd(), mustWork = FALSE)
  if (basename(cwd) == "r") {
    root_dir <- normalizePath(file.path(cwd, ".."), mustWork = FALSE)
  } else {
    root_dir <- cwd
  }
}

old_wd <- getwd()
setwd(root_dir)
on.exit(setwd(old_wd), add = TRUE)

if (!file.exists(file.path("outputs", "tables", "urban_systems_timeseries.csv"))) {
  py <- Sys.which("python3")
  if (py == "") py <- Sys.which("python")
  if (py != "") {
    system2(py, file.path("python", "urban_systems_model.py"))
  }
}

source(file.path("r", "urban_systems_diagnostics.R"), local = TRUE)
cat("\nAll base R urban systems workflows completed.\n")
